Hdf5 1.14.5

blowekamp opened the pull request:
<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the ITK repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->

hjmjohnson reviewed (changes requested):
There is a WIP commit that indicates this PR is not yet ready for review, so I made it a draft.

dzenanz commented on Modules/ThirdParty/HDF5/src/CMakeLists.txt:
```diff
@@ -64,6 +63,23 @@ set(HDF_TEST_EXPRESS OFF CACHE INTERNAL "Control testing framework (0-3)")
 set(BUILD_USER_DEFINED_LIBS OFF CACHE INTERNAL "Build user defined libs")
 
+set (HDF5_DIMENSION_SCALES_NEW_REF OFF CACHE INTERNAL "Use new-style references with dimension scale APIs" )
+set (HDF5_EXTERNAL_LIB_PREFIX "itk" CACHE INTERNAL "Use prefix for custom library naming.")
```
@jcfr this has a potential to interact with https://github.com/Slicer/ITK/commit/aacaefd6366bb09e09faefd173d4d3797f386ad0 and other [Slicer customizations](https://github.com/Slicer/ITK/commits/slicer-v5.4.0-2024-05-16-311b706/).

dzenanz commented on Modules/ThirdParty/HDF5/src/CMakeLists.txt:
```diff
@@ -64,6 +63,23 @@ set(HDF_TEST_EXPRESS OFF CACHE INTERNAL "Control testing framework (0-3)")
+set( HDF5_ENABLE_MAP_API              OFF CACHE INTERNAL "Build the map API")
+set( HDF5_ENABLE_WARNINGS_AS_ERRORS  OFF CACHE INTERNAL "Interpret some warnings as errors")
+#set( HDF5_MINGW_STATIC_GCC_LIBS       OFF CACHE INTERNAL "")
+set( HDF5_MSVC_NAMING_CONVENTION      OFF CACHE INTERNAL "Use MSVC Naming conventions for Shared Libraries")
```
Does this matter for Windows/MSVC?

dzenanz commented on Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt:
```diff
@@ -1,5 +1,11 @@
 cmake_minimum_required (VERSION 3.12)
 project (HDF5_SRC C)
+# ITK --start
+set(MANGLE_PREFIX itk)
```
Should we make `MANGLE_PREFIX` an ITK top-level option?

dzenanz reviewed (approved):
Generally looks good

blowekamp commented:
The following command can be run to show the local changes in the third party library:
`
$ git rev-list upstream-HDF5 -n 1
83f14e7ecba1f91b4152f692f54e56306ba3e182
$git diff "upstream-HDF5" HEAD:./Modules/ThirdParty/HDF5/src/itkhdf5/
`

There are only a couple changes that appear to be needed locally still:
bdc523419f3fe180addc122136a6e6912209b47c @hjmjohnson Did this get contributed upstream?

blowekamp commented:
The CI for the builds are passing 🎉 

However ghost flow checker as a couple issues:
- missing newline at the end of file in src/H5Odtype.c.
   - Is there a way to suppress this for this one file, or do we ignore ghost flow here?
- hdf.icns with size 326012 bytes (318.37 KiB) which is greater than the maximum size 300000 bytes (292.97 KiB). 
  - This is just an icon file, not code. It's used with CPack, may not be reachable. Maybe it should be "rm"'d in the UpdateFromUpstream script?
 - H5Ppublic.h with size 404419 bytes (394.94 KiB)
   - So this needs to be added to the .gitattributes file. Does that change need to occur before the commit where it's added?

It's going to be difficult to change the initial sub-tree commit and redo the merge. I want to make sure I have all the necessary changes done in the correct order before going through that process again.

hjmjohnson commented:
Ensure that #4716 modifications are also included in this PR.

dzenanz commented:
It is better to make changes to `.gitattributes` and the update script before before running it. We don't have to fix all the ghostflow warnings.

blowekamp commented on Modules/ThirdParty/HDF5/.gitattributes:
```diff
@@ -2,3 +2,5 @@ src/itkhdf5/src/H5C.c        hooks-max-size=500000
 src/itkhdf5/src/H5Shyper.c   hooks-max-size=600000
 src/itkhdf5/src/H5Tconv.c    hooks-max-size=500000
+src/itkhdf5/src/H5Ppublic.h  hooks-max-size=500000
+src/H5Odtype.c               no-lf-at-eof
```
This attribute did not work to remove the Ghostflow checker error. The error comes from the sub tree. Likely the attribute should be in the subtree not the main tree. I don't think it's worth the effort to redo the merge again to address this.

blowekamp commented on Modules/ThirdParty/HDF5/src/itkhdf5/src/H5Apublic.h:
```diff
@@ -1009,6 +1037,42 @@ H5_DLL herr_t H5Awrite(hid_t attr_id, hid_t type_id, const void *buf);
+/* (Must be defined _after_ the function prototype) */
+/* (And must only defined when included in application code, not the library) */
+#ifndef H5A_MODULE
+#define H5Acreate_async(...)         HD5_MANGLE_PREFIX##_H5Acreate_async(__FILE__, __func__, __LINE__, __VA_ARGS__)
```
@thewtex @dzenanz These definitions like this are problematic for the name mangling. The current redefinitions work, but produce warnings. Each header ( 8-ish) has a "?_MODULE" conditional so that the definitional are applied externally/publicly but not applied when building the library.

My current thought is to add #undef before each of these secondary definitions. But this will be a lot of local modifications. Any other suggestions?

dzenanz replied:
I haven't looked at HDF5 mangling in a long time. Do what you think is best.

thewtex replied:
If possible, pre-processor updates are added to `itk_hdf5_mangle.h.in` and this is included first. If this approach can be adapted to apply for this issue (?), this will keep the modifications in one place outside the upstream code, which would ease maintenance.

blowekamp replied:
Unfortunately, you can not redefine a definition. It's rather unfortunate that the "internal" function name and the public CPP define are the same symbol and I can not thing of a way to name mangle it without the code modifications in the latest update.

blowekamp commented:
This PR is done as far as I am currently aware now.

thewtex commented:
Were the cross-platform name mangling symbol checks described in itk_hdf5_mangle.h.in run?

blowekamp commented:
> Were the cross-platform name mangling symbol checks described in itk_hdf5_mangle.h.in run?

I ran the process on OSX arm and linux arm.

thewtex commented:
We should cover Windows amd64, Linux amd64. There are often new symbols with an update.

hjmjohnson commented:
@thewtex Do you have a windows box for doing the mangling?

dzenanz commented:
I have a Windows computer, but I expect to be quite busy on Monday and Tuesday.

blowekamp commented:
Merging before the final symbols are updated should not cause problems with the CIs or regular builds. The only issue would be for applications used ITK master and are integrating with other untangled HDF5 libraries.

thewtex commented:
> @thewtex Do you have a windows box for doing the mangling?

I am traveling now, but I could take a look next week.



> Merging before the final symbols are updated should not cause problems with the CIs or regular builds. The only issue would be for applications used ITK master and are integrating with other untangled HDF5 libraries.

These symbols have caused crashes in Slicer, and there are likely similar applications out there. I would prefer not to merge until all the platforms have been checked.

CC @jcfr

seanm commented:
> Merging before the final symbols are updated should not cause problems with the CIs or regular builds. The only issue would be for applications used ITK master and are integrating with other untangled HDF5 libraries.

Some of my nightly ITK builds on cdash build against VTK too, so they might catch any duplicate symbols.

Thanks for working on this.  I took a stab at a smaller update [here](https://github.com/InsightSoftwareConsortium/ITK/pull/4716) but got stuck on various cmake stuff I did not understand.

I could clone this PR locally and try building my own app & test suite against it...

seanm commented:
With `ITK_USE_SYSTEM_ZLIB=ON` (my typical) it fails to build with:

```
CMake Error at Modules/ThirdParty/HDF5/src/itkhdf5/src/CMakeLists.txt:1216 (get_target_property):
  get_target_property() called with non-existent target
  "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib/libz.tbd".
```

seanm commented:
Turning `ITK_USE_SYSTEM_ZLIB` off, I was able to build my app!  My app uses VTK & ITK and I built with ASan too. No link errors. No unit test failures.

So for me, this is in good shape, just needs a fix for `ITK_USE_SYSTEM_ZLIB=ON`.

dzenanz commented:
I extracted symbols. I am not sure whether extraction scripts need an update - there are many symbols with extra closed parenthesis. [win_hdf5_mangle.zip](https://github.com/user-attachments/files/17835813/win_hdf5_mangle.zip)

Also, build in Debug mode fails - I included the build log. `C:\Misc\ITK-patches2\Modules\ThirdParty\HDF5\src\itkhdf5\src\H5make_libsettings.c` indeed does not exist.

dzenanz commented:
Building in `RelWithDebInfo` mode succeeds, after which building in `Debug` mode also succeeds.

dzenanz commented:
[win_hdf5_mangle.zip](https://github.com/user-attachments/files/17836058/win_hdf5_mangle.zip) I now also added the (large) intermediate outputs. Brad, can you merge it in?

blowekamp commented:
> I extracted symbols. I am not sure whether extraction scripts need an update - there are many symbols with extra closed parenthesis.

It looks like there are a bunch go C++ symbols ( HD5 namespace ). The scripts may not work with those. I believe we were creating the C++ library before but not mangling the symbols. We should check if ITK is using the C++ libraries. 

Also I updated the docs to indicate that the mangling should be disabled before extracting the symbols. The defines already have the "itk_" prefix on them. I think I can remove those.

blowekamp commented:
@thewtex As far as I know has been completed for this PR. Including integration of the windows symbol mangling.

Shall we merge?

thewtex commented:
@blowekamp thanks!

Shall we merge? 👍

seanm commented:
This is exciting!  A lot of remaining UBSan issues were due to HDF... I should give that another try now...
